host/xtest/pkcs11_1000.c

```diff
@@ -3115,3 +3115,534 @@ static void xtest_pkcs11_test_1013(ADBG_Case_t *c)
+	CK_FLAGS rw_session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	CK_FLAGS ro_session_flags = CKF_SERIAL_SESSION;
+	CK_OBJECT_HANDLE obj_hdl = CK_INVALID_HANDLE;
+	CK_OBJECT_HANDLE obj_hdl_ro = CK_INVALID_HANDLE;
+	CK_OBJECT_HANDLE obj_hdl_cp = CK_INVALID_HANDLE;
+	const char *label = "Dummy Objects";
+	CK_ATTRIBUTE secret_key_template[] = {
+		{ CKA_CLASS, &(CK_OBJECT_CLASS){CKO_SECRET_KEY},
+						sizeof(CK_OBJECT_CLASS) },
+		{ CKA_KEY_TYPE,	&(CK_KEY_TYPE){CKK_AES}, sizeof(CK_KEY_TYPE) },
+		{ CKA_TOKEN, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_PRIVATE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_MODIFIABLE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_COPYABLE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_DESTROYABLE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+	};
+	CK_BBOOL g_token = CK_FALSE;
+	CK_BBOOL g_private = CK_FALSE;
+	CK_BBOOL g_modify = CK_FALSE;
+	CK_BBOOL g_copy = CK_FALSE;
+	CK_BBOOL g_destroy = CK_FALSE;
+	CK_BBOOL g_extract = CK_FALSE;
+	CK_BBOOL g_sensitive = CK_FALSE;
+	CK_BBOOL g_nextract = CK_FALSE;
+	CK_BBOOL g_asensitive = CK_FALSE;
+	CK_BBOOL g_local =  CK_FALSE;
+	CK_ATTRIBUTE get_template[] = {
+		{ CKA_TOKEN, &g_token, sizeof(CK_BBOOL) },
+		{ CKA_PRIVATE, &g_private, sizeof(CK_BBOOL) },
+		{ CKA_MODIFIABLE, &g_modify, sizeof(CK_BBOOL) },
+		{ CKA_COPYABLE, &g_copy, sizeof(CK_BBOOL) },
+		{ CKA_DESTROYABLE, &g_destroy, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &g_extract, sizeof(CK_BBOOL) },
+		{ CKA_SENSITIVE, &g_sensitive, sizeof(CK_BBOOL) },
+		{ CKA_NEVER_EXTRACTABLE, &g_nextract, sizeof(CK_BBOOL) },
+		{ CKA_ALWAYS_SENSITIVE, &g_asensitive, sizeof(CK_BBOOL) },
+		{ CKA_LOCAL, &g_local, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE copy_template[] = {
+		{ CKA_TOKEN, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_MODIFIABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_SENSITIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE copy_template_inv[] = {
+		{ CKA_APPLICATION, (CK_UTF8CHAR_PTR)label, sizeof(label) },
+	};
+	CK_ATTRIBUTE copy_template_priv[] = {
+		{ CKA_PRIVATE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE set_template[] = {
+		{ CKA_COPYABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+	};
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return;
+
+	rv = init_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	rv = init_user_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	/* Open a RW session */
+	rv = C_OpenSession(slot, rw_session_flags, NULL, 0, &rw_session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	/* Open a RO session */
+	rv = C_OpenSession(slot, ro_session_flags, NULL, 0, &ro_session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
```

Comment:
`goto out;`
or new label as we have `rw_session` session open.

Reply:
Thanks. I will change it to goto out and add a check there
if (ro_session != CK_INVALID_HANDLE)
                ADBG_EXPECT_CK_OK(c, C_CloseSession(ro_session));